src/compiler/scala/tools/nsc/ast/parser/SyntaxAnalyzer.scala

```diff
@@ -91,7 +95,21 @@ abstract class SyntaxAnalyzer extends SubComponent with Parsers with MarkupParse
     override val checkable = false
     override val keepsTypeParams = false
 
-    def apply(unit: CompilationUnit): Unit = {
+    override def run() {
+      echoPhaseSummary(this)
+
+      val threads = scala.util.Properties.envOrElse("PARSER_THREADS", "4").toInt
+
+      if (threads == 0) {
+        currentRun.units.foreach(applyPhase)
+      } else {
+        implicit val ec = ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(threads))
+        val future = Future.traverse(currentRun.units)(unit => Future(applyPhase(unit)))
+        Await.ready(future, Duration.Inf)
```

Comment:
you need to consume the result, not just wait for it to be ready

You will also need to unbuffer messages to reporter etc to make the compile deterministic